File: platypus-js-web-client/src/main/java/com/eas/grid/columns/Column.java
package com.eas.grid.columns;

import com.eas.core.HasPublished;
import com.eas.core.Utils;
import com.eas.grid.PathComparator;
import com.eas.ui.HasJsValue;
import com.eas.ui.PublishedCell;
import com.eas.ui.Widget;
import com.eas.widgets.WidgetsUtils;
import com.eas.widgets.boxes.BooleanDecoratorField;
import com.eas.widgets.boxes.CheckBox;
import com.eas.widgets.boxes.RadioButton;
import com.google.gwt.core.client.JavaScriptObject;
import com.google.gwt.dom.client.Document;
import com.google.gwt.dom.client.InputElement;
import com.google.gwt.dom.client.StyleElement;
import com.google.gwt.dom.client.TableCellElement;
import com.google.gwt.dom.client.TableColElement;
import com.google.gwt.user.client.ui.HasText;
import com.eas.core.Logger;
import com.eas.core.XElement;
import com.eas.grid.Grid;
import com.eas.grid.HeaderView;
import com.google.gwt.dom.client.BrowserEvents;
import com.google.gwt.dom.client.NativeEvent;
import java.util.List;

// TODO: Check tree expandable cell decorations like left paddnig according to deepness and plus / minus icon and open / closed folder icons
public class Column implements HasPublished {

    private TableColElement element = Document.get().createColElement();
    private String radioGroup = "group-" + Document.get().createUniqueId();
    private StyleElement columnRule = Document.get().createStyleElement();
    protected String field;
    protected String sortField;
    protected Widget editor;
    /**
     * Minimum column width while resizing by a user.
     */
    protected double minWidth = 15;
    /**
     * Maximum column width while resizing by a user.
     */
    protected double maxWidth = Integer.MAX_VALUE;
    protected double width = 75;
    protected boolean readonly;
    protected boolean visible = true;
    protected boolean selectOnly;
    protected boolean sortable;
    protected int indent = 24;
    protected PathComparator comparator;
    private HeaderView header;
    protected JavaScriptObject published;
    protected JavaScriptObject onRender;
    protected JavaScriptObject onSelect;
    protected Grid grid;

    public Column() {
        super();
    }

    public TableColElement getElement() {
        return element;
    }

    public HeaderView getHeader() {
        return header;
    }

    public void setHeader(HeaderView header) {
        this.header = header;
    }

    public Grid getGrid() {
        return grid;
    }

    public void setGrid(Grid grid) {
        this.grid = grid;
    }

    public StyleElement getColumnRule() {
        return columnRule;
    }

    public PathComparator getComparator() {
        return comparator;
    }

    public void sort() {
        sort(true);
    }

    public void sort(boolean fireEvent) {
        comparator = new PathComparator(sortField != null && !sortField.isEmpty() ? sortField : field, true);
        if (fireEvent) {
            grid.sort();
        }
    }

    public void sortDesc() {
        sortDesc(true);
    }

    public void sortDesc(boolean fireEvent) {
        comparator = new PathComparator(sortField != null && !sortField.isEmpty() ? sortField : field, false);
        if (fireEvent) {
            grid.sort();
        }
    }

    public void unsort() {
        unsort(true);
    }

    public void unsort(boolean fireEvent) {
        comparator = null;
        if (fireEvent) {
            grid.sort();
        }
    }

    public String getField() {
        return field;
    }

    public void setField(String aValue) {
        if (field == null ? aValue != null : !field.equals(aValue)) {
            field = aValue;
        }
    }

    public String getSortField() {
        return sortField;
    }

    public void setSortField(String aValue) {
        if (sortField == null ? aValue != null : !sortField.equals(aValue)) {
            sortField = aValue;
        }
    }

    public double getMinWidth() {
        return minWidth;
    }

    public void setMinWidth(double aValue) {
        minWidth = aValue;
    }

    public double getMaxWidth() {
        return maxWidth;
    }

    public void setMaxWidth(double aValue) {
        maxWidth = aValue;
    }

    public Object getValue(JavaScriptObject anElement) {
        if (anElement != null && field != null && !field.isEmpty()) {
            return Utils.getPathData(anElement, field);
        } else {
            return null;
        }
    }

    public void update(JavaScriptObject anElement, JavaScriptObject value) {
        if (anElement != null && field != null && !field.isEmpty() && !readonly) {
            Utils.setPathData(anElement, field, value);
        }
    }

    public void render(int viewIndex, JavaScriptObject dataRow, TableCellElement viewCell) {
        List<JavaScriptObject> path = grid.buildPathTo(dataRow);
        int padding = indent * path.size();
        Object value = getValue(dataRow);
        String text;
        if (value == null) {
            text = null; // No native rendering for null values
        } else {
            text = String.valueOf(value);
            if (editor != null) {
                ((HasJsValue) editor).setJsValue(value);
                if (editor instanceof BooleanDecoratorField || editor instanceof CheckBox) {
                    InputElement checkbox = Document.get().createCheckInputElement();
                    if (Boolean.TRUE.equals(value)) {
                        checkbox.setChecked(true);
                    } else {
                        checkbox.setChecked(false);
                        text = "";
                    }
                    viewCell.appendChild(checkbox);
                    checkbox.<XElement>cast().addEventListener(BrowserEvents.CHANGE, new XElement.NativeHandler() {
                        @Override
                        public void on(NativeEvent evt) {
                            update(dataRow, !Boolean.TRUE.equals(value)); //  !!!value
                        }

                    });
                } else if (editor instanceof RadioButton) {
                    InputElement radiobutton = Document.get().createRadioInputElement(radioGroup);
                    if (Boolean.TRUE.equals(value)) {
                        radiobutton.setChecked(true);
                    } else {
                        radiobutton.setChecked(false);
                        text = "";
                    }
                    viewCell.appendChild(radiobutton);
                    radiobutton.<XElement>cast().addEventListener(BrowserEvents.CHANGE, new XElement.NativeHandler() {
                        @Override
                        public void on(NativeEvent evt) {
                            if (Boolean.FALSE.equals(value)) {
                                update(dataRow, true);
                            }
                        }

                    });
                } else {
                    text = ((HasText) editor).getText();
                    viewCell.setInnerText(text);
                }
            }
        }
        if (onRender != null || grid.getOnRender() != null) { // User's rendering for all values, including null
            try {
                // TODO: Check if abstract 'cell' object is needed at all  
                PublishedCell cellToRender = WidgetsUtils.calcGridPublishedCell(published, onRender != null ? onRender : grid.getOnRender(), dataRow, field, text, viewCell, viewIndex, null);
                if (cellToRender != null) {
                    if (cellToRender.getDisplayCallback() == null) {
                        cellToRender.setDisplayCallback(new Runnable() {
                            @Override
                            public void run() {
                                cellToRender.styleToElement(viewCell);
                                viewCell.setInnerText(cellToRender.getDisplay());
                            }
                        });
                    }
                    cellToRender.styleToElement(viewCell);
                    viewCell.setInnerText(cellToRender.getDisplay());
                }
            } catch (Exception ex) {
                Logger.severe(ex);
            }
        }
    }

    public boolean isVisible() {
        return visible;
    }

    public void setVisible(boolean aValue) {
        if (visible != aValue) {
            visible = aValue;
            if (visible) {
                columnRule.setInnerHTML("");
            } else {
                columnRule.setInnerHTML("display: none");
            }
        }
    }

    public double getDesignedWidth() {
        return width;
    }

    public double getWidth() {
        return width;
    }

    public void setWidth(double aValue) {
        width = aValue;
    }

    public boolean isReadonly() {
        return readonly;
    }

    public void setReadonly(boolean aValue) {
        readonly = aValue;
    }

    public boolean isSortable() {
        return sortable;
    }

    public void setSortable(boolean aValue) {
        sortable = aValue;
    }

    public boolean isSelectOnly() {
        return selectOnly;
    }

    public void setSelectOnly(boolean aValue) {
        selectOnly = aValue;
    }

    public JavaScriptObject getOnRender() {
        return onRender;
    }

    public void setOnRender(JavaScriptObject aValue) {
        if (onRender != aValue) {
            onRender = aValue;
        }
    }

    public JavaScriptObject getOnSelect() {
        return onSelect;
    }

    public void setOnSelect(JavaScriptObject aValue) {
        if (onSelect != aValue) {
            onSelect = aValue;
        }
    }

    public Widget getEditor() {
        return editor;
    }

    public void setEditor(Widget aEditor) {
        if (editor != aEditor) {
            editor = aEditor;
        }
    }

    /*
    public void setEditor(ValueDecoratorField aEditor) {
        if (editor != aEditor) {
            if (editor != null) {
                editor.setOnRender(null);
                editor.setOnSelect(null);
                if (editor instanceof DropDownListDecoratorField) {
                    ((DropDownListDecoratorField) editor).setOnRedraw(null);
                }
            }
            editor = aEditor;
            if (editor != null) {
                editor.setOnRender(onRender);
                editor.setOnSelect(onSelect);
                if (editor instanceof DropDownListDecoratorField) {
                    ((DropDownListDecoratorField) editor).setOnRedraw(new Runnable() {

                        @Override
                        public void run() {
                            enqueueGridRedraw();
                        }

                    });
                }
            }
            if (editor instanceof BooleanDecoratorField) {
                ((TreeExpandableCell<JavaScriptObject, Object>) getCell()).setCell(new CheckBoxCell() {
                    @Override
                    public void onBrowserEvent(com.google.gwt.cell.client.Cell.Context context, Element parent, Object value, NativeEvent event, ValueUpdater<Object> valueUpdater) {
                        String type = event.getType();
                        if (BrowserEvents.CHANGE.equals(type) && (readonly || !grid.isEditable())) {
                            InputElement input = parent.<XElement>cast().firstChildByTagName("input").cast();
                            boolean checked = input.isChecked();
                            input.setChecked(!checked);
                        } else {
                            super.onBrowserEvent(context, parent, value, event, valueUpdater);
                        }
                    }

                    @Override
                    public void render(com.google.gwt.cell.client.Cell.Context context, Object aValue, SafeHtmlBuilder sb) {
                        try {
                            ValueDecoratorField modelEditor = getEditor();
                            aValue = modelEditor.convert(Utils.toJava(aValue));
                            super.render(context, aValue, sb);
                        } catch (Exception ex) {
                            Logger.severe(ex);
                        }
                    }

                });
            } else {
                ((TreeExpandableCell<JavaScriptObject, Object>) getCell()).setCell(new RenderedEditorCell<Object>(editor) {
                    @Override
                    protected void renderCell(com.google.gwt.cell.client.Cell.Context context, Object value, SafeHtmlBuilder sb) {
                        try {
                            ValueDecoratorField modelEditor = getEditor();
                            value = modelEditor.convert(Utils.toJava(value));
                            super.renderCell(context, value, sb);
                        } catch (Exception ex) {
                            Logger.severe(ex);
                        }
                    }

                    @Override
                    public void startEditing(com.google.gwt.cell.client.Cell.Context context, Element aBoxPositionTemplate, Element aBoxParent, Object value, ValueUpdater<Object> valueUpdater,
                            Runnable onEditorClose) {
                        try {
                            ValueDecoratorField modelEditor = getEditor();
                            value = modelEditor.convert(Utils.toJava(value));
                            grid.setActiveEditor(getEditor());
                            super.startEditing(context, aBoxPositionTemplate, aBoxParent, value, valueUpdater, onEditorClose);
                        } catch (Exception ex) {
                            Logger.severe(ex);
                        }
                    }
                });
                RenderedEditorCell<Object> cell = (RenderedEditorCell<Object>) getTargetCell();
                cell.setReadonly(new CellHasReadonly() {

                    @Override
                    public boolean isReadonly() {
                        return readonly || !grid.isEditable();
                    }

                });
                cell.setOnEditorClose(new RenderedEditorCell.EditorCloser() {
                    @Override
                    public void closed(Element aTable) {
                        grid.setActiveEditor(null);
                        final GridSection<?> toFocus = GridSection.getInstance(aTable);
                        if (toFocus != null) {
                            Scheduler.get().scheduleDeferred(new Scheduler.ScheduledCommand() {

                                @Override
                                public void execute() {
                                    toFocus.setFocus(true);
                                }
                            });
                        }
                    }
                });
                cell.setRenderer(new CellRenderer<Object>() {

                    @Override
                    public boolean render(Context context, String aId, Object value, SafeHtmlBuilder sb) {
                        JavaScriptObject onRender = Column.this.getOnRender() != null ? Column.this.getOnRender() : Column.this.getGrid().getOnRender();
                        if (onRender != null) {
                            ValueDecoratorField modelEditor = getEditor();
                            value = modelEditor.convert(Utils.toJava(value));
                            String display = null;
                            Object oldValue = ((HasValue<Object>) editor).getValue();
                            ((HasValue<Object>) editor).setValue(value);
                            try {
                                display = ((HasText) editor).getText();
                            } finally {
                                // We have to take care about old value because
                                // of edited cell while grid rendering.
                                ((HasValue<Object>) editor).setValue(oldValue);
                            }
                            SafeHtmlBuilder lsb = new SafeHtmlBuilder();
                            PublishedCell cellToRender = calcContextPublishedCell(Column.this.getPublished(), onRender, context, Column.this.getField(), display);
                            if (cellToRender != null) {
                                if (cellToRender.getDisplay() != null) {
                                    display = cellToRender.getDisplay();
                                }
                            }

                            if (display == null) {
                                lsb.append(SafeHtmlUtils.fromTrustedString("&#160;"));
                            } else {
                                lsb.append(SafeHtmlUtils.fromString(display));
                            }
                            grid.complementPublishedStyle(cellToRender);
                            String decorId = renderDecorated(lsb, aId, cellToRender, sb);
                            if (cellToRender != null) {
                                if (context instanceof RenderedCellContext) {
                                    ((RenderedCellContext) context).setPublishedCell(cellToRender);
                                }
                                Column.bindDisplayCallback(cellToRender, decorId);
                                Column.bindIconCallback(cellToRender, decorId);
                            }
                            return true;
                        } else {
                            return false;
                        }
                    }

                });
            }
        }
    }

    public static PublishedCell calcContextPublishedCell(JavaScriptObject aThis, JavaScriptObject aOnRender, com.google.gwt.cell.client.Cell.Context context, String aField, String aDisplay) {
        if (aOnRender != null) {
            Object key = context.getKey();
            JavaScriptObject renderedElement = key instanceof JavaScriptObject ? (JavaScriptObject) key : null;
            if (renderedElement != null) {
                Object data = aField != null && !aField.isEmpty() ? Utils.getPathData(renderedElement, aField) : null;
                PublishedCell cell = WidgetsUtils.publishCell(data, aDisplay);
                JsArrayMixed args = JavaScriptObject.createArray().cast();
                args.push(EventsPublisher.publishOnRenderEvent(aThis, null, null, renderedElement, cell));
                aOnRender.<JsObject>cast().apply(aThis, args);
                return cell;
            }
        }
        return null;
    }

    protected static void bindDisplayCallback(final PublishedCell aCell, final String aTargetElementId) {
        aCell.setDisplayCallback(new Runnable() {
            @Override
            public void run() {
                Element identifiedTextSection = Document.get().getElementById(aTargetElementId);
                if (identifiedTextSection != null) {
                    aCell.styleToElementBackgroundToTd(identifiedTextSection);
                    String toRender = aCell.getDisplay();
                    if (toRender == null) {
                        toRender = "&#160;";
                    }
                    identifiedTextSection.setInnerSafeHtml(SafeHtmlUtils.fromTrustedString(toRender));
                }
            }
        });
    }

    protected static void bindIconCallback(final PublishedCell aCell, final String aTargetElementId) {
        aCell.setIconCallback(new Runnable() {

            @Override
            public void run() {
                Element identifiedTextSection = Document.get().getElementById(aTargetElementId);
                if (identifiedTextSection != null) {
                    ImageElement iconSection = (ImageElement) identifiedTextSection.getPreviousSiblingElement();
                    if (iconSection != null) {
                        if (aCell.getIcon() != null) {
                            iconSection.setSrc(aCell.getIcon().getSafeUri().asString());
                        } else {
                            iconSection.setSrc(null);
                        }
                    }
                }
            }

        });
    }
     */
    @Override
    public JavaScriptObject getPublished() {
        return published;
    }

    @Override
    public void setPublished(JavaScriptObject aValue) {
        if (published != aValue) {
            published = aValue;
        }
    }

}

File: platypus-js-web-client/src/app/forms/grid/column.js
/* global Infinity */
define([
    '../../id',
    '../../ui',
    '../bound',
    '../path-comparator'
], function (
        Id,
        Ui,
        Bound,
        PathComparator
        ) {
// TODO: Check tree expandable cell decorations like left paddnig according to deepness and plus / minus icon and open / closed folder icons
    function Column(node) {
        var self = this;
        var cols = []; // dom 'col' elements for header,frozen,body and footer sections of the grid
        var columnRule = document.createElement('style');
        var columnStyleName = 'p-grid-column-' + Id.generate();
        var field = null;
        var sortField = null;
        var editor = null;
        /**
         * Minimum column width while resizing by a user.
         */
        var minWidth = 15;
        /**
         * Maximum column width while resizing by a user.
         */
        var maxWidth = Infinity;
        var width = 75;
        var padding = 0;
        var readonly = false;
        var visible = true;
        var sortable = true;
        var comparator; // PathComparator
        var headers = []; // multiple instances of NodeView
        var onRender;
        var onSelect;
        var grid;

        function regenerateColStyle() {
            columnRule.innerHTML = '.' + columnStyleName + '{' +
                    (visible ? '' : 'display: none;') +
                    (width == null || width == Infinity ? '' : 'width: ' + (width + padding) + 'px;') +
                    (minWidth == null || minWidth == Infinity ? '' : 'min-width: ' + minWidth + 'px;') +
                    (maxWidth == null || maxWidth == Infinity ? '' : 'max-width: ' + maxWidth + 'px;') +
                    '}';
        }
        regenerateColStyle();

        Object.defineProperty(this, 'styleName', {
            get: function () {
                return columnStyleName;
            }
        });

        function addCol() {
            var col = document.createElement('col');
            cols.push(col);
            col.className = columnStyleName;
            return col;
        }

        Object.defineProperty(this, 'addCol', {
            get: function () {
                return addCol;
            }
        });
        /**
         * Multiple 'col' elements for the single column, because of grid sections.
         */
        Object.defineProperty(this, 'elements', {
            get: function () {
                return cols;
            }
        });

        /**
         * Multiple 'headers' for the single column, because of splitted column nodes.
         */
        Object.defineProperty(this, 'headers', {
            get: function () {
                return headers;
            }
        });
        /**
         * Typically, we need only leaf column's header.
         * Leaf nodes' columns can have only single header, by nature.
         */
        Object.defineProperty(this, 'header', {
            get: function () {
                return headers.length === 1 ? headers[0] : null;
            }
        });
        Object.defineProperty(this, 'grid', {
            get: function () {
                return grid;
            },
            set: function (aValue) {
                grid = aValue;
            }
        });

        Object.defineProperty(this, 'node', {
            get: function () {
                return node;
            }
        });

        Object.defineProperty(this, 'columnRule', {
            get: function () {
                return columnRule;
            }
        });

        Object.defineProperty(this, 'comparator', {
            get: function () {
                return comparator;
            },
            set: function (aValue) {
                comparator = aValue;
            }
        });

        function sort(fireEvent) {
            if (arguments.length < 1)
                fireEvent = true;
            comparator = new PathComparator(sortField ? sortField : field, true);
            if (fireEvent) {
                grid.addSortedColumn(self);
            }
        }
        Object.defineProperty(this, 'sort', {
            get: function () {
                return sort;
            }
        });

        function sortDesc(fireEvent) {
            if (arguments.length < 1)
                fireEvent = true;
            comparator = new PathComparator(sortField ? sortField : field, false);
            if (fireEvent) {
                grid.addSortedColumn(self);
            }
        }
        Object.defineProperty(this, 'sortDesc', {
            get: function () {
                return sortDesc;
            }
        });

        function unsort(fireEvent) {
            if (arguments.length < 1)
                fireEvent = true;
            comparator = null;
            if (fireEvent) {
                grid.removeSortedColumn(self);
            }
        }
        Object.defineProperty(this, 'unsort', {
            get: function () {
                return unsort;
            }
        });

        Object.defineProperty(this, 'field', {
            get: function () {
                return field;
            },
            set: function (aValue) {
                if (field !== aValue) {
                    field = aValue;
                }
            }
        });

        Object.defineProperty(this, 'sortField', {
            get: function () {
                return sortField;
            },
            set: function (aValue) {
                if (sortField !== aValue) {
                    sortField = aValue;
                }
            }
        });

        Object.defineProperty(this, 'width', {
            get: function () {
                return width;
            },
            set: function (aValue) {
                if (width !== aValue) {
                    width = aValue;
                    regenerateColStyle();
                    if (grid) {
                        grid.updateSectionsWidth();
                    }
                }
            }
        });

        Object.defineProperty(this, 'padding', {
            get: function () {
                return padding;
            },
            set: function (aValue) {
                if (aValue != null && padding !== aValue) {
                    padding = aValue;
                    regenerateColStyle();
                    if (grid) {
                        grid.updateSectionsWidth();
                    }
                }
            }
        });

        Object.defineProperty(this, 'minWidth', {
            configurable: true,
            get: function () {
                return minWidth;
            },
            set: function (aValue) {
                if (minWidth !== aValue) {
                    minWidth = aValue;
                    regenerateColStyle();
                }
            }
        });

        Object.defineProperty(this, 'maxWidth', {
            configurable: true,
            get: function () {
                return maxWidth;
            },
            set: function (aValue) {
                if (maxWidth !== aValue) {
                    maxWidth = aValue;
                    regenerateColStyle();
                }
            }
        });

        function getValue(aItem) {
            if (aItem && field) {
                return Bound.getPathData(aItem, field);
            } else {
                return null;
            }
        }
        Object.defineProperty(this, 'getValue', {
            configurable: true,
            get: function () {
                return getValue;
            }
        });

        function setValue(anElement, value) {
            if (anElement && field && !readonly) {
                Bound.setPathData(anElement, field, value);
            }
        }
        Object.defineProperty(this, 'setValue', {
            configurable: true,
            get: function () {
                return setValue;
            }
        });

        function render(viewIndex, dataRow, viewCell) {
            if (grid.treeIndicatorColumn === self) {
                var padding = grid.indent * grid.depthOf(dataRow);
                viewCell.style.paddingLeft = padding > 0 ? padding + 'px' : '';
                if (!grid.isLeaf(dataRow)) {
                    viewCell.classList.add(grid.expanded(dataRow) ? 'p-grid-cell-expanded' : 'p-grid-cell-collapsed');
                }
                Ui.on(viewCell, Ui.Events.MOUSEDOWN, function (event) {
                    if (event.button === 0) {
                        var rect = viewCell.getBoundingClientRect();
                        if (event.clientX > rect.left + padding - grid.indent &&
                                event.clientX <= rect.left + padding) {
                            event.stopPropagation();
                            grid.toggle(dataRow);
                        }
                    }
                });
            }
            var value = getValue(dataRow);
            if (value == null) { // null == undefined, null !== undefined
                viewCell.innerTHML = ''; // No native rendering for null values
            } else if (typeof (value) === 'boolean') {
                var checkbox = document.createElement('input');
                checkbox.type = 'checkbox';
                checkbox.checked = !!value;
                viewCell.appendChild(checkbox);
            } else {
                var html;
                if (editor) {
                    editor.value = value;
                    html = editor.text;
                } else if (value instanceof Date) {
                    html = value.toJSON();
                } else {
                    html = value + '';
                }
                viewCell.innerHTML = html;
            }
            // User's rendering for all values, including null
            if (onRender || grid.onRender) {
                var handler = onRender ? onRender : grid.onRender;
                handler.call(self, dataRow, viewCell, viewIndex, html);
            }
        }

        Object.defineProperty(this, 'render', {
            configurable: true,
            get: function () {
                return render;
            }
        });
        Object.defineProperty(this, 'visible', {
            get: function () {
                return visible;
            },
            set: function (aValue) {
                if (visible !== aValue) {
                    visible = aValue;
                    regenerateColStyle();
                    if (grid) {
                        grid.applyColumnsNodes();
                    }
                }
            }
        });

        Object.defineProperty(this, 'readonly', {
            get: function () {
                return readonly;
            },
            set: function (aValue) {
                readonly = aValue;
            }
        });

        Object.defineProperty(this, 'sortable', {
            get: function () {
                return sortable;
            },
            set: function (aValue) {
                sortable = aValue;
            }
        });

        Object.defineProperty(this, 'onRender', {
            get: function () {
                return onRender;
            },
            set: function (aValue) {
                if (onRender !== aValue) {
                    onRender = aValue;
                }
            }
        });

        Object.defineProperty(this, 'onSelect', {
            get: function () {
                return editor ? editor.onSelect : null;
            },
            set: function (aValue) {
                if (onSelect !== aValue && editor) {
                    editor.onSelect = aValue;
                }
            }
        });

        Object.defineProperty(this, 'editor', {
            get: function () {
                return editor;
            },
            set: function (aValue) {
                if (editor !== aValue) {
                    editor = aValue;
                }
            }
        });
    }
    return Column;
});